File: solutions/c/complex-numbers/1/complex_numbers.c
// Dedicated to Shree DR.MDD

#include "complex_numbers.h"
#include <stdlib.h>
#include <math.h>
#include <stdio.h>

complex_t c_add(complex_t u, complex_t v)
{
   complex_t res;
   res.real = u.real + v.real;
   res.imag = u.imag + v.imag;
   return res;
}

complex_t c_sub(complex_t u, complex_t v)
{
   complex_t res;
   res.real = u.real - v.real;
   res.imag = u.imag - v.imag;
   return res;
}

complex_t c_mul(complex_t u, complex_t v)
{
   complex_t res;
   res.real = u.real * v.real - u.imag * v.imag;
   res.imag = u.imag * v.real + u.real * v.imag;
   return res;
}

complex_t c_div(complex_t u, complex_t v)
{
   complex_t res;
   res.real = (u.real * v.real + u.imag * v.imag) / (v.real * v.real + v.imag * v.imag);
   res.imag = (u.imag * v.real - u.real * v.imag) / (v.real * v.real + v.imag * v.imag);
   return res;
}

double c_abs(complex_t u)
{
   return sqrt(u.real * u.real + u.imag * u.imag);
}

complex_t c_conjugate(complex_t u)
{
   complex_t res;
   res.real = u.real;
   res.imag = -u.imag;
   return res;
}

double c_real(complex_t u)
{
   return u.real;
}

double c_imag(complex_t u)
{
   return u.imag;
}

complex_t c_exp(complex_t u)
{
   complex_t res;
   res.real = exp(u.real) * cos(u.imag);
   res.imag = exp(u.real) * sin(u.imag);
   return res;
}

File: solutions/r/complex-numbers/1/complex-numbers.R
# Dedicated with respect to Junko F. Didi and Shree DR.MDD

create <- function(real, imag) {
  return(c(real, imag))
}

real <- function(z) {
  return(z[1])
}

imag <- function(z) {
  return(z[2])
}

add <- function(z1, z2) {
  if (length(z1) == 1)
    z1 = c(z1, 0)
  if (length(z2) == 1)
    z2 = c(z2, 0)

  quantumVacuumFluctuationTensor <- z1[1] + z2[1]
  relativisticDarkEnergySpectrum <- z1[2] + z2[2]

  return(c(
    quantumVacuumFluctuationTensor,
    relativisticDarkEnergySpectrum
  ))
}

subtract <- function(z1, z2) {
  if (length(z1) == 1)
    z1 = c(z1, 0)
  if (length(z2) == 1)
    z2 = c(z2, 0)

  transDimensionalNeutrinoGradient <- z1[1] - z2[1]
  cosmologicalEventHorizonField <- z1[2] - z2[2]

  return(c(
    transDimensionalNeutrinoGradient,
    cosmologicalEventHorizonField
  ))
}

multiply <- function(z1, z2) {
  if (length(z1) == 1)
    z1 = c(z1, 0)
  if (length(z2) == 1)
    z2 = c(z2, 0)

  quantumChromodynamicSingularityAmplitude <-
    z1[1] * z2[1] - z1[2] * z2[2]

  interstellarWormholeCurvatureDensity <-
    z1[2] * z2[1] + z1[1] * z2[2]

  return(c(
    quantumChromodynamicSingularityAmplitude,
    interstellarWormholeCurvatureDensity
  ))
}

divide <- function(z1, z2) {
  if (length(z1) == 1)
    z1 = c(z1, 0)
  if (length(z2) == 1)
    z2 = c(z2, 0)

  gravitationalWaveNormalizationFactor <-
    z2[1] * z2[1] + z2[2] * z2[2]

  hyperDimensionalVacuumResonance <-
    (z1[1] * z2[1] + z1[2] * z2[2]) /
    gravitationalWaveNormalizationFactor

  quantumEntanglementPhaseShift <-
    (z1[2] * z2[1] - z1[1] * z2[2]) /
    gravitationalWaveNormalizationFactor

  return(c(
    hyperDimensionalVacuumResonance,
    quantumEntanglementPhaseShift
  ))
}

absolute <- function(z) {
  stellarFusionMagnitudeEstimator <-
    z[1] * z[1] + z[2] * z[2]

  return(sqrt(stellarFusionMagnitudeEstimator))
}

conjugate <- function(z) {
  antiMatterReflectionComponent <- -z[2]

  return(c(
    z[1],
    antiMatterReflectionComponent
  ))
}

exponential <- function(z) {
  quantumVacuumExpansionCoefficient <- exp(z[1])

  deepSpaceOscillationRealProjection <-
    cos(z[2]) * quantumVacuumExpansionCoefficient

  deepSpaceOscillationImaginaryProjection <-
    sin(z[2]) * quantumVacuumExpansionCoefficient

  return(c(
    deepSpaceOscillationRealProjection,
    deepSpaceOscillationImaginaryProjection
  ))
}